compiler/rustc_mir/src/transform/early_otherwise_branch.rs

```diff
@@ -239,13 +240,13 @@ impl<'a, 'tcx> Helper<'a, 'tcx> {
         if is_switch(terminator) {
             let this_bb_discr_info = self.find_switch_discriminant_info(bb, terminator)?;
 
-            // the types of the two adts matched on have to be equalfor this optimization to apply
-            if discr_info.type_adt_matched_on != this_bb_discr_info.type_adt_matched_on {
-                trace!(
-                    "NO: types do not match. LHS: {:?}, RHS: {:?}",
-                    discr_info.type_adt_matched_on,
-                    this_bb_discr_info.type_adt_matched_on
-                );
+            // The layouts of the two ADTs have to be equal for this optimization to apply
```

Comment:
Please explain this a bit more. It will be very surprising for future readers. It's not actually the layouts we care about, but the discriminant ids corresponding to the variants that we're looking for

Reply:
We do have all the information at compile-time, maybe we should use it to figure out whether the two variants of different types that we're looking at have the same discriminant as returned by the `Discriminant` rvalue. That would be even less conservative

Reply:
I'm a bit unsure what part of Layout (https://doc.rust-lang.org/nightly/nightly-rustc/rustc_target/abi/struct.Layout.html) actually needs to match between the two types for us to say that this optimization is valid.

Reply:
I don't think you need to check anything in the layout, you can check each type for the discriminant values directly: https://github.com/rust-lang/rust/blob/c2bc344eb23d8c1d18e803b3f1e631cf99926fbb/compiler/rustc_mir/src/interpret/operand.rs#L657-L666

If you do this, please deduplicate your logic with the logic in const eval

Reply:
Hmm, in what cases is it actually even necessary to check the type of the discriminants? We already check in `find_discriminant_switch_pairing` that the discriminant value of both switches match. See (of which the first check is redundant with the second) https://github.com/rust-lang/rust/blob/9b77a6a20021cb8783111c04e9260963f15b550e/compiler/rustc_mir/src/transform/early_otherwise_branch.rs#L258-L282

So this code https://godbolt.org/z/vW3j4h can still be optimized as both the lhs and lhs of the match are of the same discriminant value (1).

Do you see any problems with simply removing the layout check?

Reply:
so... to recap. If the discriminants of the two variants being checked are the same *and* a very specific value, then we apply the optimization. This sounds to me like we can generalize the optimization even further, since almost the same thing happens with

```rust
if x == 42 && y == 42 {
    99
} else {
    69
}
```

I wasn't able to produce the same MIR as we get with the match, but if we can come up with user-code that produces the same code without using a match, then we should think about generalizing it.

Anyway, back to this PR. Yea, we need neither the redundant value check nor the type/layout equality check, as these don't actually prevent miscompilations as far as I can tell

Reply:
For this PR: The newest commit removes the layout check and the redundant value check. So this should be ready for a review again.

For a future PR: Yeah true, your snippet does seem like something awfully lot like something this optimization could handle.